nav2_core/include/nav2_core/global_planner.hpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Samsung Research America
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NAV2_CORE_GLOBAL_PLANNER_H_
+#define NAV2_CORE_GLOBAL_PLANNER_H_
+
+#include <string>
+#include "rclcpp/rclcpp.h"
+#include "nav2_costmap_2d/nav2_costmap_2d.hpp"
+#include "tf2_ros/Buffer.h"
+#include "nav2_msgs/msg/path.h"
+#include "geometry_msgs/msg/pose_stamped.h"
+
+namespace nav2_core
+{
+
+/**
+ * @class GlobalPlanner
+ * @brief Abstract interface for global planners to adhere to with pluginlib
+ */
+class GlobalPlanner
+{
+public:
+  /**
+   * @brief Virtual destructor
+   */
+  virtual ~GlobalPlanner() {}
+
+  /**
+   * @param  parent pointer to user's node
+   * @param  name The name of this planner
+   * @param  tf A pointer to a TF buffer
+   * @param  costmap_ros A pointer to the costmap
+   */
+  virtual void configure(const rclcpp::Node * parent,
+    const std::string & name, tf2_ros::Buffer * tf,
+    nav2_costmap_2d::Costmap2DROS * costmap_ros) = 0;
+
+  /**
+   * @brief Method to cleanup resources used on shutdown.
+   */
+  virtual void cleanup() = 0;
+
+  /**
+   * @brief Method to active planner and any threads involved in execution.
+   */
+  virtual void activate() = 0;
+
+  /**
+   * @brief Method to deactive planner and any threads involved in execution.
+   */
+  virtual void deactivate() = 0;
+
+  /**
```

Comment:
Just for completeness, to match the lifecycle state machine, probably need a virtual shutdown() method also.

Reply:
I didn't add it only because when I looked at other nodes they didnt seem to do anything in `on_error` and `on_shutdown`, should I add `on_error` too?

Reply:
I added `shutdown` let me know thoughts on `error`

Reply:
I know that @mjeronimo had thought to create a `LifecycleHelperInterface` to derive from to align with the lifecycle state machine, in particular differentiating that these helper classes wouldn't need to implement an `on_error` or `on_shutdown`. I'm more inclined to think we could just derive from the `LifecycleNodeInterface`  itself and just override the ones that we care about. Perhaps we could do that throughout the stack where appropriate.

https://github.com/ros2/rclcpp/blob/master/rclcpp_lifecycle/include/rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp

Reply:
From glancing, it looks like `LifecycleHelperInterface` is just virtualizing all the necessary items for lifecycle implementation, I dont know that we want to power that down to the plugin layer all that structure. I wanted to pare it down to the minimum set of useful things that can be easily understood without a knowledge of lifecycle: configuration, pausing, activating (and now stopping). 

Does `LifecycleNodeInterface` make a node? We want to give it a node not create one and have additional overhead.

Reply:
> From glancing, it looks like `LifecycleHelperInterface` is just virtualizing all the necessary items for lifecycle implementation, I dont know that we want to power that down to the plugin layer all that structure. I wanted to pare it down to the minimum set of useful things that can be easily understood without a knowledge of lifecycle: configuration, pausing, activating (and now stopping). If we were to make plugins conform to all the lifecycle APIs, then yes that makes sense to use.
> 
The `LifecycleHelperInterface` just pairs down the interface from `LifecycleNodeInterface` to not include the `on_error` and `on_shutdown` and is a pure virtual interface. I'm not sure it's really necessary as opposed to using the full interface. I think it would make sense to conform the plugins in our stack. We can always have default behavior to transition the nodes automatically (and hence the plugins). 

> Does `LifecycleNodeInterface` make a node? We want to give it a node not create one and have additional overhead, its not necessarily clear to me from that file what the differences are beyond the `=0` to force the user to create it.

I linked the class so it would be clear, the `LifecycleNodeInterface` doesn't create nor receive a node, it just introduces the transition calls `on_activate`, `on_configure`, etc.

Reply:
Ok, if you have a strong feeling that you'd like to power down the lifecycle API to the plugin layer, I could get in line with that. Of the two, I agree `LifecycleNodeInterface` is the better of the two for the reason that its nearly the same as `LifecycleHelperInterface` and its not some specialized header for this project. 

Let me know and I'll make those changes.

Reply:
Since we are already somewhat trying to match the lifecycle state machine, I see it as we might as well use the interface. I'm working on cleaning up some of the system-level lifecycle transitions, so I think this may be a good pattern to align the plugin model.

Reply:
OK. Are you OK with me using the `LifecycleNodeInterface` or do you  want me to derive from `LifecycleHelperInterface`? `LifecycleNodeInterface` not forcing the user to implement the functions is more in line with my thinking

Reply:
Yes, I'm okay with `LifecycleNodeInterface`

Reply:
I'm also not opposed to creating something like a `LifecyclePluginInterace` if that makes more sense in this context

Reply:
One challenge I see is that the `on_configure` call is generic and thus you'd have to introduce the planner specific objects elsewhere. I'd ideally like to abstract all the transition interfaces so you can simply derive for all plugins, though.

Reply:
I did not think of that. Can we do something else here then? Have the plugins implement `on_configure` as `configure` and all the `on_...` as `...`s so we can introduce our own things to pass into the plugins? They could still be functionally treated as lifecycle states in the server's use of the plugins 
```
void on_configure()
{
  for plugin in plugins:
    plugin->configure(stuff, thing, object);
}
```

Reply:
Deriving from Lifecycle not allowing all the items required would be not great. I suppose rather than in `configure` we could give those pointers in the plugin's constructor, however that would then encourage bad behavior by setting things up there vs on the `configure` state signal